controllers/service/service_controller.go

```diff
@@ -99,16 +108,102 @@ func (r *serviceReconciler) reconcile(ctx context.Context, req ctrl.Request) err
 	if err := r.k8sClient.Get(ctx, req.NamespacedName, svc); err != nil {
 		return client.IgnoreNotFound(err)
 	}
+	if err := r.allocatedService(ctx, svc); err != nil {
+		return err
+	}
 	stack, lb, backendSGRequired, err := r.buildModel(ctx, svc)
 	if err != nil {
 		return err
 	}
-	if lb == nil {
-		return r.cleanupLoadBalancerResources(ctx, svc, stack)
+	if lb == nil || !svc.DeletionTimestamp.IsZero() {
+		return r.cleanupLoadBalancerResources(ctx, svc, stack, lb == nil)
 	}
 	return r.reconcileLoadBalancerResources(ctx, svc, stack, lb, backendSGRequired)
 }
 
+// allocatedService allocates a stack to a service so that it can share load balancer resources with other services.
+func (r *serviceReconciler) allocatedService(ctx context.Context, svc *corev1.Service) error {
+	if !r.initialized {
+		r.lock.Lock()
+		var serviceList corev1.ServiceList
+		if err := r.k8sClient.List(ctx, &serviceList); err != nil {
+			return err
+		}
+		for _, svc := range serviceList.Items {
+			if svc.Annotations[service.LoadBalancerStackKey] != "" {
+				r.allocatedServices[svc.Annotations[service.LoadBalancerStackKey]] = map[int]bool{}
```

Comment:
nit: if you only want to track existence, this is more efficient.
```suggestion
				r.allocatedServices[svc.Annotations[service.LoadBalancerStackKey]] = map[int]struct{}{}
```

You could also make the maps `map[int32]struct{}{}` to avoid the `int(port.NodePort)` everywhere.

Reply:
👍 addressed